File: host/src/hidpin/device.py
"""Talking to a hidpin device over USB HID."""

from __future__ import annotations

import os
import random
from dataclasses import dataclass
from types import TracebackType

from hidpin import _hidraw, protocol
from hidpin.protocol import (
    DeviceInfo,
    PinConfig,
    PinConfigReport,
    PinMode,
    PinSetting,
    Reason,
    Result,
    StatusReport,
)

VENDOR_ID = 0x1209
PRODUCT_ID = 0x0001
USAGE_PAGE = 0xFF00
USAGE = 0x01

REPORT_LEN = 1 + protocol.REPORT_PAYLOAD_LEN


class HidpinError(Exception):
    """Base class for every error raised by this library."""


class DeviceNotFound(HidpinError):
    pass


class ProtocolVersionError(HidpinError):
    def __init__(self, device_version: int) -> None:
        super().__init__(
            f"device speaks protocol version {device_version}, this library speaks {protocol.PROTOCOL_VERSION}"
        )
        self.device_version = device_version


class PinConfigRejected(HidpinError):
    def __init__(self, result: Result | int, gpio: int, *, local: bool = False) -> None:
        message = protocol.RESULT_MESSAGES.get(result, f"result {int(result)}")
        if gpio != protocol.RESULT_GPIO_NONE:
            message = f"{message} (GPIO{gpio})"
        where = "ピン設定が不正です (デバイスには送っていません)" if local else "デバイスがピン設定を拒否しました"
        super().__init__(f"{where}: {message}")
        self.result = result
        self.gpio = gpio
        self.local = local


class PinConfigConflict(HidpinError):
    """Another process changed the pin configuration at the same time (PROTOCOL.md 6.3)."""

    def __init__(self, expected: int, seen: int) -> None:
        super().__init__(f"別のプロセスがピン設定を変更しました (request_id {expected} を送り {seen} が返りました)")
        self.expected = expected
        self.seen = seen


@dataclass(frozen=True)
class DeviceEntry:
    """One hidpin device found on the host."""

    path: bytes
    serial: str
    manufacturer: str
    product: str


def _import_hid():
    try:
        import hid
    except ImportError as exc:  # pragma: no cover - depends on the host environment
        raise HidpinError(
            "hidapi が見つかりません。'pip install hidapi' を実行するか、Linux では hidraw を使ってください"
        ) from exc
    return hid


def _default_backend():
    """Linux では hidraw を使い、それ以外では hidapi を使う。

    HIDPIN_BACKEND=hidraw / hidapi で明示的に選べる。hidapi の PyPI ホイールは libusb
    バックエンドで作られており、カーネルドライバを奪えないと開けないことがあるため、
    Linux では hidraw を既定にしている。
    """
    choice = os.environ.get("HIDPIN_BACKEND", "auto").lower()
    if choice not in ("auto", "hidraw", "hidapi"):
        raise HidpinError(f"HIDPIN_BACKEND には hidraw か hidapi を指定してください: '{choice}'")
    if choice in ("auto", "hidraw") and _hidraw.available():
        return _hidraw
    if choice == "hidraw":
        raise HidpinError("hidraw が使えません (/sys/class/hidraw が見つかりません)")
    return _import_hid()


def _is_hidpin(item: dict) -> bool:
    usage_page = item.get("usage_page") or 0
    # Some backends report 0 when they cannot parse the report descriptor.
    return usage_page in (0, USAGE_PAGE)


def find_devices(backend=None) -> list[DeviceEntry]:
    """Lists connected hidpin devices."""
    hid = backend or _default_backend()
    entries = []
    for item in hid.enumerate(VENDOR_ID, PRODUCT_ID):
        if not _is_hidpin(item):
            continue
        entries.append(
            DeviceEntry(
                path=item.get("path", b""),
                serial=item.get("serial_number") or "",
                manufacturer=item.get("manufacturer_string") or "",
                product=item.get("product_string") or "",
            )
        )
    return sorted(entries, key=lambda entry: entry.serial)


class Device:
    """A connected hidpin device.

    Not thread safe: use one instance from one thread.
    """

    def __init__(self, handle, *, serial: str = "") -> None:
        self._handle = handle
        self.serial = serial
        self._info: DeviceInfo | None = None
        self._config: PinConfig | None = None
        self._polarity: dict[int, bool] = {}
        self._last_request_id = 0
        self._last_seq: int | None = None
        self.missed_reports = 0

    @classmethod
    def open(cls, serial: str | None = None, *, path: bytes | None = None, backend=None) -> "Device":
        hid = backend or _default_backend()
        if path is None:
            entries = find_devices(backend=hid)
            if serial is not None:
                entries = [entry for entry in entries if entry.serial == serial]
            if not entries:
                raise DeviceNotFound("hidpin デバイスが見つかりません" if serial is None else f"シリアル番号 {serial} のデバイスが見つかりません")
            if len(entries) > 1 and serial is None:
                found = ", ".join(entry.serial for entry in entries)
                raise DeviceNotFound(f"複数のデバイスが接続されています。--serial で指定してください: {found}")
            path = entries[0].path
            serial = entries[0].serial
        handle = hid.device()
        try:
            handle.open_path(path)
        except OSError as exc:
            raise HidpinError(
                f"デバイスを開けませんでした ({exc})。"
                "権限が足りない場合は udev ルールを入れて、ボードを挿し直してください: "
                "sudo cp udev/70-hidpin.rules /etc/udev/rules.d/ && "
                "sudo udevadm control --reload-rules && sudo udevadm trigger"
            ) from exc
        device = cls(handle, serial=serial or "")
        device.info  # verifies the protocol version before anything else
        return device

    def close(self) -> None:
        self._handle.close()

    def __enter__(self) -> "Device":
        return self

    def __exit__(self, exc_type: type[BaseException] | None, exc: BaseException | None, tb: TracebackType | None) -> None:
        self.close()

    # --- reports -----------------------------------------------------------------

    def _get_feature(self, report_id: int) -> bytes:
        data = bytes(self._handle.get_feature_report(report_id, REPORT_LEN))
        if len(data) < REPORT_LEN:
            raise HidpinError(f"Feature レポート {report_id} の長さが {len(data)} バイトしかありません")
        if data[0] != report_id:
            raise HidpinError(f"Feature レポート {report_id} を要求しましたが {data[0]} が返りました")
        return data[1:REPORT_LEN]

    def _send_feature(self, report_id: int, payload: bytes) -> None:
        self._handle.send_feature_report(bytes([report_id]) + payload)

    @property
    def info(self) -> DeviceInfo:
        """Device information; read once and cached (PROTOCOL.md 5)."""
        if self._info is None:
            info = protocol.decode_device_info(self._get_feature(protocol.REPORT_ID_DEVICE_INFO))
            if info.protocol_version != protocol.PROTOCOL_VERSION:
                raise ProtocolVersionError(info.protocol_version)
            self._info = info
        return self._info

    def get_pin_config(self) -> PinConfigReport:
        report = protocol.decode_pin_config(self._get_feature(protocol.REPORT_ID_PIN_CONFIG))
        self._config = report.config
        return report

    @property
    def pin_config(self) -> PinConfig:
        if self._config is None:
            self.get_pin_config()
        assert self._config is not None
        return self._config

    def _next_request_id(self) -> int:
        request_id = random.randint(1, 255)
        while request_id == self._last_request_id:
            request_id = random.randint(1, 255)
        self._last_request_id = request_id
        return request_id

    def set_pin_config(self, config: PinConfig, *, request_id: int | None = None) -> PinConfigReport:
        """Writes a pin configuration and verifies it was applied (PROTOCOL.md 6.3)."""
        result, gpio = protocol.validate_pin_config(config, self.info.available)
        if result != Result.OK:
            raise PinConfigRejected(result, gpio, local=True)

        request_id = request_id if request_id is not None else self._next_request_id()
        self._send_feature(protocol.REPORT_ID_PIN_CONFIG, protocol.encode_pin_config_set(config, request_id))

        report = self.get_pin_config()
        if report.request_id != request_id:
            raise PinConfigConflict(request_id, report.request_id)
        if report.result != Result.OK:
            raise PinConfigRejected(report.result, report.result_gpio)
        if report.config != config:
            raise HidpinError("デバイスが適用したピン設定が、送った内容と一致しません")
        return report

    def update_pins(self, settings: dict[int, PinSetting]) -> PinConfigReport:
        """Changes some pins, keeping every other pin as it is."""
        return self.set_pin_config(self.get_pin_config().config.with_pins(settings))

    def request_status(self) -> StatusReport:
        """Asks for the current state with Get_Report(Input) (PROTOCOL.md 4.6)."""
        data = bytes(self._handle.get_input_report(protocol.REPORT_ID_STATUS, REPORT_LEN))
        if len(data) < REPORT_LEN or data[0] != protocol.REPORT_ID_STATUS:
            raise HidpinError("状態通知の取得に失敗しました")
        return protocol.decode_status(data[1:REPORT_LEN])

    def read_status(self, timeout_ms: int = 1000) -> StatusReport | None:
        """Waits for the next status report; returns None on timeout."""
        data = self._handle.read(REPORT_LEN, timeout_ms)
        if not data:
            return None
        data = bytes(data)
        if data[0] != protocol.REPORT_ID_STATUS:
            raise HidpinError(f"予期しないレポート ID {data[0]} を受け取りました")
        report = protocol.decode_status(data[1:REPORT_LEN])
        self._track_seq(report)
        return report

    def _track_seq(self, report: StatusReport) -> None:
        if report.reason & Reason.HOST_REQUEST:
            return
        if self._last_seq is not None:
            missed = (report.seq - self._last_seq - 1) & 0xFFFF
            self.missed_reports += missed
        self._last_seq = report.seq

    # --- outputs -----------------------------------------------------------------

    def write_outputs(self, mask: int, value: int) -> None:
        payload = protocol.encode_output(mask, value)
        self._handle.write(bytes([protocol.REPORT_ID_OUTPUT]) + payload)

    def set_outputs(self, levels: dict[int, bool]) -> None:
        """Drives output pins; GPIOs that are not output pins are ignored by the device."""
        mask = 0
        value = 0
        for gpio, level in levels.items():
            mask |= 1 << gpio
            if level:
                value |= 1 << gpio
        self.write_outputs(mask, value)

    # --- polarity ----------------------------------------------------------------

    def set_polarity(self, gpio: int, active_low: bool) -> None:
        """Overrides which level counts as ON for one pin."""
        self._polarity[gpio] = active_low

    def is_active_low(self, gpio: int, config: PinConfig | None = None) -> bool:
        """LOW means ON for pins with a pull-up, unless the caller said otherwise."""
        if gpio in self._polarity:
            return self._polarity[gpio]
        config = config if config is not None else self.pin_config
        return config[gpio].mode == PinMode.PULLUP

    def on_off(self, report: StatusReport, config: PinConfig | None = None) -> dict[int, bool]:
        """Interprets the pin levels of monitored pins as ON/OFF."""
        config = config if config is not None else self.pin_config
        states = {}
        for gpio in protocol.mask_to_gpios(report.monitored):
            level = report.level(gpio)
            states[gpio] = (not level) if self.is_active_low(gpio, config) else level
        return states

File: host-go/hidpin/device.go
package hidpin

import (
	"errors"
	"fmt"
	"math/rand/v2"
	"os"
	"strconv"
	"strings"
	"time"
)

// USB identifiers. The defaults are the pid.codes test PID; override them with HIDPIN_VID and
// HIDPIN_PID when the firmware was built with other ids.
const (
	DefaultVendorID  uint16 = 0x1209
	DefaultProductID uint16 = 0x0001
)

// USBIDs returns the vendor and product id to look for.
func USBIDs() (vendorID, productID uint16, err error) {
	if vendorID, err = envID("HIDPIN_VID", DefaultVendorID); err != nil {
		return 0, 0, err
	}
	if productID, err = envID("HIDPIN_PID", DefaultProductID); err != nil {
		return 0, 0, err
	}
	return vendorID, productID, nil
}

func envID(name string, fallback uint16) (uint16, error) {
	raw := strings.TrimSpace(os.Getenv(name))
	if raw == "" {
		return fallback, nil
	}
	value, err := strconv.ParseUint(raw, 0, 64)
	if err != nil {
		return 0, fmt.Errorf("%s is not a 16-bit number: '%s'", name, raw)
	}
	if value > 0xFFFF {
		return 0, fmt.Errorf("%s is out of range for 16 bits: '%s'", name, raw)
	}
	return uint16(value), nil
}

// Transport carries reports to and from one device. Every report starts with its report ID.
// The Linux hidraw backend implements it; tests use a fake.
type Transport interface {
	// Read waits for one input report. It returns 0 and no error when the timeout expires; a
	// negative timeout waits forever.
	Read(buf []byte, timeout time.Duration) (int, error)
	// Write sends an output report.
	Write(report []byte) (int, error)
	// GetFeatureReport reads a feature report; buf[0] selects the report ID.
	GetFeatureReport(buf []byte) (int, error)
	// SendFeatureReport writes a feature report.
	SendFeatureReport(report []byte) (int, error)
	// GetInputReport reads an input report on the control pipe; buf[0] selects the report ID.
	GetInputReport(buf []byte) (int, error)
	Close() error
}

// Entry is one hidpin device found on the host.
type Entry struct {
	Path         string
	Serial       string
	Manufacturer string
	Product      string
}

// ErrNotFound is returned when no matching device is connected.
var ErrNotFound = errors.New("no hidpin device found")

// ErrUnsupportedPlatform is returned on operating systems without a backend.
var ErrUnsupportedPlatform = errors.New("hidpin supports Linux (hidraw) only")

// ProtocolVersionError means the device speaks a protocol version this library does not.
type ProtocolVersionError struct {
	DeviceVersion uint8
}

func (e *ProtocolVersionError) Error() string {
	return fmt.Sprintf("device speaks protocol version %d, this library speaks %d", e.DeviceVersion, ProtocolVersion)
}

// PinConfigRejectedError means a pin configuration was invalid. Local is true when the library
// caught it before anything was sent to the device.
type PinConfigRejectedError struct {
	Result Result
	GPIO   uint8
	Local  bool
}

func (e *PinConfigRejectedError) Error() string {
	where := "the device rejected the pin configuration"
	if e.Local {
		where = "the pin configuration is invalid (nothing was sent to the device)"
	}
	message := e.Result.Message()
	if e.GPIO != ResultGPIONone {
		message = fmt.Sprintf("%s (GPIO%d)", message, e.GPIO)
	}
	return where + ": " + message
}

// PinConfigConflictError means another process changed the pin configuration at the same time
// (PROTOCOL.md 6.3).
type PinConfigConflictError struct {
	Expected uint8
	Seen     uint8
}

func (e *PinConfigConflictError) Error() string {
	return fmt.Sprintf("another process changed the pin configuration (sent request_id %d, read back %d)", e.Expected, e.Seen)
}

// Open opens the connected device with the given serial number, or the only connected device
// when serial is empty.
func Open(serial string) (*Device, error) {
	entries, err := FindDevices()
	if err != nil {
		return nil, err
	}
	if serial != "" {
		matching := entries[:0]
		for _, e := range entries {
			if e.Serial == serial {
				matching = append(matching, e)
			}
		}
		entries = matching
	}
	switch {
	case len(entries) == 0 && serial == "":
		return nil, ErrNotFound
	case len(entries) == 0:
		return nil, fmt.Errorf("%w: no device with serial number %s", ErrNotFound, serial)
	case len(entries) > 1 && serial == "":
		serials := make([]string, len(entries))
		for i, e := range entries {
			serials[i] = e.Serial
		}
		return nil, fmt.Errorf("several devices are connected; pick one with --serial: %s", strings.Join(serials, ", "))
	}
	return OpenPath(entries[0].Path, entries[0].Serial)
}

// Device is a connected hidpin device. It is not safe for concurrent use.
type Device struct {
	transport Transport
	// Serial is the USB serial number the device was opened by.
	Serial string

	info          *DeviceInfo
	config        *PinConfig
	polarity      map[int]bool
	lastRequestID uint8
	lastSeq       uint16
	haveSeq       bool
	missed        int
}

// NewDevice wraps a transport; the protocol version is checked on the first call to Info.
func NewDevice(transport Transport, serial string) *Device {
	return &Device{transport: transport, Serial: serial, polarity: map[int]bool{}}
}

// Close closes the transport.
func (d *Device) Close() error {
	return d.transport.Close()
}

func (d *Device) getFeature(reportID uint8) ([]byte, error) {
	buf := make([]byte, ReportLen)
	buf[0] = reportID
	n, err := d.transport.GetFeatureReport(buf)
	if err != nil {
		return nil, fmt.Errorf("reading feature report %d: %w", reportID, err)
	}
	if n < ReportLen {
		return nil, fmt.Errorf("feature report %d is only %d bytes long", reportID, n)
	}
	if buf[0] != reportID {
		return nil, fmt.Errorf("asked for feature report %d but got report %d", reportID, buf[0])
	}
	return buf[1:ReportLen], nil
}

// Info returns the device information, reading it once and checking the protocol version.
func (d *Device) Info() (DeviceInfo, error) {
	if d.info != nil {
		return *d.info, nil
	}
	payload, err := d.getFeature(ReportIDDeviceInfo)
	if err != nil {
		return DeviceInfo{}, err
	}
	info, err := DecodeDeviceInfo(payload)
	if err != nil {
		return DeviceInfo{}, err
	}
	if info.ProtocolVersion != ProtocolVersion {
		return DeviceInfo{}, &ProtocolVersionError{DeviceVersion: info.ProtocolVersion}
	}
	d.info = &info
	return info, nil
}

// GetPinConfig reads the pin configuration together with the result of the last write.
func (d *Device) GetPinConfig() (PinConfigReport, error) {
	payload, err := d.getFeature(ReportIDPinConfig)
	if err != nil {
		return PinConfigReport{}, err
	}
	report, err := DecodePinConfig(payload)
	if err != nil {
		return PinConfigReport{}, err
	}
	d.config = &report.Config
	return report, nil
}

// PinConfig returns the pin configuration, reading it only if it has not been read yet.
func (d *Device) PinConfig() (PinConfig, error) {
	if d.config == nil {
		if _, err := d.GetPinConfig(); err != nil {
			return PinConfig{}, err
		}
	}
	return *d.config, nil
}

func (d *Device) nextRequestID() uint8 {
	for {
		id := uint8(rand.IntN(255) + 1)
		if id != d.lastRequestID {
			d.lastRequestID = id
			return id
		}
	}
}

// SetPinConfig writes a pin configuration and verifies it was applied (PROTOCOL.md 6.3).
func (d *Device) SetPinConfig(config PinConfig) (PinConfigReport, error) {
	info, err := d.Info()
	if err != nil {
		return PinConfigReport{}, err
	}
	if result, gpio := ValidatePinConfig(config, info.Available); result != ResultOK {
		return PinConfigReport{}, &PinConfigRejectedError{Result: result, GPIO: gpio, Local: true}
	}

	requestID := d.nextRequestID()
	payload, err := EncodePinConfigSet(config, requestID)
	if err != nil {
		return PinConfigReport{}, err
	}
	if _, err := d.transport.SendFeatureReport(append([]byte{ReportIDPinConfig}, payload...)); err != nil {
		return PinConfigReport{}, fmt.Errorf("writing the pin configuration: %w", err)
	}

	report, err := d.GetPinConfig()
	if err != nil {
		return PinConfigReport{}, err
	}
	if report.RequestID != requestID {
		return PinConfigReport{}, &PinConfigConflictError{Expected: requestID, Seen: report.RequestID}
	}
	if report.Result != ResultOK {
		return PinConfigReport{}, &PinConfigRejectedError{Result: report.Result, GPIO: report.ResultGPIO}
	}
	if report.Config != config {
		return PinConfigReport{}, errors.New("the configuration the device applied differs from the one that was sent")
	}
	return report, nil
}

// UpdatePins changes some pins and keeps every other pin as it is.
func (d *Device) UpdatePins(settings map[int]PinSetting) (PinConfigReport, error) {
	current, err := d.GetPinConfig()
	if err != nil {
		return PinConfigReport{}, err
	}
	config, err := current.Config.WithPins(settings)
	if err != nil {
		return PinConfigReport{}, err
	}
	return d.SetPinConfig(config)
}

// RequestStatus asks for the current state with Get_Report(Input) (PROTOCOL.md 4.6).
func (d *Device) RequestStatus() (Status, error) {
	buf := make([]byte, ReportLen)
	buf[0] = ReportIDStatus
	n, err := d.transport.GetInputReport(buf)
	if err != nil {
		return Status{}, fmt.Errorf("could not read the status report: %w", err)
	}
	if n < ReportLen || buf[0] != ReportIDStatus {
		return Status{}, errors.New("could not read the status report")
	}
	return DecodeStatus(buf[1:ReportLen])
}

// ReadStatus waits for the next status report. It returns false when the timeout expires.
func (d *Device) ReadStatus(timeout time.Duration) (Status, bool, error) {
	buf := make([]byte, ReportLen)
	n, err := d.transport.Read(buf, timeout)
	if err != nil {
		return Status{}, false, err
	}
	if n == 0 {
		return Status{}, false, nil
	}
	if buf[0] != ReportIDStatus {
		return Status{}, false, fmt.Errorf("received an unexpected report ID %d", buf[0])
	}
	if n < ReportLen {
		return Status{}, false, fmt.Errorf("status report is only %d bytes long", n)
	}
	status, err := DecodeStatus(buf[1:ReportLen])
	if err != nil {
		return Status{}, false, err
	}
	d.trackSeq(status)
	return status, true, nil
}

func (d *Device) trackSeq(s Status) {
	if s.Reason&ReasonHostRequest != 0 {
		return
	}
	if d.haveSeq {
		d.missed += int(s.Seq - d.lastSeq - 1)
	}
	d.lastSeq = s.Seq
	d.haveSeq = true
}

// MissedReports counts status reports the host did not read, detected from gaps in seq.
func (d *Device) MissedReports() int {
	return d.missed
}

// WriteOutputs drives the output pins in mask to the levels in value.
func (d *Device) WriteOutputs(mask, value uint32) error {
	_, err := d.transport.Write(append([]byte{ReportIDOutput}, EncodeOutput(mask, value)...))
	return err
}

// SetOutputs drives output pins; GPIOs that are not output pins are ignored by the device.
func (d *Device) SetOutputs(levels map[int]bool) error {
	var mask, value uint32
	for gpio, high := range levels {
		if gpio < 0 || gpio >= GPIOCount {
			return fmt.Errorf("GPIO%d is out of range (0-%d)", gpio, GPIOCount-1)
		}
		mask |= 1 << gpio
		if high {
			value |= 1 << gpio
		}
	}
	return d.WriteOutputs(mask, value)
}

// SetPolarity overrides which level counts as ON for one pin.
func (d *Device) SetPolarity(gpio int, activeLow bool) {
	d.polarity[gpio] = activeLow
}

// IsActiveLow reports whether LOW means ON: true for pins with a pull-up, unless overridden.
func (d *Device) IsActiveLow(gpio int, config PinConfig) bool {
	if activeLow, ok := d.polarity[gpio]; ok {
		return activeLow
	}
	return config[gpio].Mode == ModePullUp
}

// OnOff interprets the pin levels of the monitored pins as ON/OFF.
func (d *Device) OnOff(s Status, config PinConfig) map[int]bool {
	states := map[int]bool{}
	for _, gpio := range MaskToGPIOs(s.Monitored) {
		states[gpio] = s.Level(gpio) != d.IsActiveLow(gpio, config)
	}
	return states
}
